RFC/src/RFC-0250_Covenants.md

```diff
@@ -0,0 +1,468 @@
+| `field::features`                   	| 0x04      	| output.features                   	|
+| `field::features_flags`             	| 0x05      	| output.features.flags             	|
+| `field::features_maturity`          	| 0x06      	| output.features.maturity          	|
+| `field::features_unique_id`         	| 0x07      	| output.features.unique_id         	|
+| `field::features_parent_public_key` 	| 0x08      	| output.features.parent_public_key 	|
+| `field::features_metadata`         	| 0x09      	| output.features.metadata          	|
+
+Each field tag returns a consensus encoded byte representation of the value contained in the field.
+How those bytes are interpreted depends on the covenant. For instance, `filter_fields_hashed_eq` will 
+concatenate the bytes and hash the result whereas `filter_field_int_eq` will interpret the bytes as a
+little-endian 64-byte unsigned integer.
+
+#### Set operations
+
+##### identity()
+
+The output set is returned unaltered. This rule is implicit for an empty (0 byte) covenant.
+
+```yaml
+op_byte: 0x20
+args: []
+```
+
+##### and(A, B)
+
+The intersection (\\(A \cap B\\)) of the resulting output set for covenant rules \\(A\\) and \\(B\\).
+
+```yaml
+op_byte: 0x21
+args: [Covenant, Covenant]
+```
+
+##### or(A, B)
+
+The union (\\(A \cup B\\)) of the resulting output set for covenant rules \\(A\\) and \\(B\\).
+
+```yaml
+op_byte: 0x22
+args: [Covenant, Covenant]
+```
+
+##### xor(A, B)
+
+The symmetric difference (\\(A \triangle B\\)) of the resulting output set for covenant rules \\(A\\) and \\(B\\). 
+This is, outputs that match either \\(A\\) or \\(B\\) but not both.
+
+```yaml
+op_byte: 0x23
+args: [Covenant, Covenant]
+```
+
+##### not(A)
+
+Returns the compliment of `A`. That is, all the elements of `A` are removed from the
+resultant output set.
+
+```yaml
+op_byte: 0x24
+args: [Covenant]
+```
+
+##### empty()
+
+Returns an empty set. This will always fail and, if used alone, prevents the UTXO from ever being spent.
+A more useful reason to use `empty` is in conjunction a conditional e.g. `if_else(Condition(older_rel(10)), A, empty)`
+
+```yaml
+op_byte: 0x25
+args: []
+```
+
+#### Filters
+
+##### filter_output_hash_eq(hash)
+
+Filters for a single output that matches the hash. This filter only returns zero or one outputs.
+
+```yaml
+op_byte: 0x30
+args: [Hash]
+```
+
+##### filter_fields_preserved(fields)
+
+Filter for outputs where all given fields in the input are preserved in the output.
+
+```yaml
+op_byte: 0x31
+args: [Fields]
+```
+
+##### filter_field_int_eq(field, int)
+
+Filters for outputs whose field value matches the given integer value. If the given field cannot be cast
+to an unsigned 64-bit integer, the transaction/block is rejected.
+
+```yaml
+op_byte: 0x32
+args: [Field, VarInt]
+```
+
+##### filter_fields_hashed_eq(fields, hash)
+
+```yaml
+op_byte: 0x33
+args: [Fields, VarInt]
+```
+
+##### filter_relative_height(height)
```

Comment:
This command can be done as is with TariScript? If we check if the input is valid to be spent, that is completely covered by TariScript. If we want to check if the input has a set height then its a completely different thing,

Reply:
It's mostly to support running one covenant if spent within 100 blocks and another after

```
or(
     and(filter_relative_height(100), A),
     B   
)
```